naming/dns_resolver.go

```diff
@@ -0,0 +1,284 @@
+type AddressType uint8
+
+const (
+	// Backend indicates the server is a backend server.
+	Backend AddressType = iota
+	// GRPCLB indicates the server is a grpclb load balancer.
+	GRPCLB
+)
+
+// AddrMetadataGRPCLB contains the information the name resolver for grpclb should provide. The
+// name resolver used by the grpclb balancer is required to provide this type of metadata in
+// its address updates.
+type AddrMetadataGRPCLB struct {
+	// AddrType is the type of server (grpc load balancer or backend).
+	AddrType AddressType
+	// ServerName is the name of the grpc load balancer. Used for authentication.
+	ServerName string
+}
+
+// compileUpdate compares the old resolved addresses and newly resolved addresses,
+// and generates an update list
+func (w *dnsWatcher) compileUpdate(newAddrs []*Update) []*Update {
+	update := make(map[Update]bool)
+	for _, u := range newAddrs {
+		update[*u] = true
+	}
+	for _, u := range w.curAddrs {
+		if _, ok := update[*u]; ok {
+			delete(update, *u)
+			continue
+		}
+		update[Update{Addr: u.Addr, Op: Delete, Metadata: u.Metadata}] = true
+	}
+	res := make([]*Update, 0, len(update))
+	for k := range update {
+		tmp := k
+		res = append(res, &tmp)
+	}
+	return res
+}
+
+func (w *dnsWatcher) lookupSRV() []*Update {
+	var newAddrs []*Update
+	_, srvs, err := lookupSRV(w.ctx, "grpclb", "tcp", w.host)
+	if err != nil {
+		grpclog.Infof("grpc: failed dns SRV record lookup due to %v.\n", err)
+		return nil
+	}
+	for _, s := range srvs {
+		lbAddrs, err := lookupHost(w.ctx, s.Target)
+		if err != nil {
+			grpclog.Warningf("grpc: failed load banlacer address dns lookup due to %v.\n", err)
+			continue
+		}
+		for _, a := range lbAddrs {
+			a, ok := formatIP(a)
+			if !ok {
+				grpclog.Errorf("grpc: failed IP parsing due to %v.\n", err)
+				continue
+			}
+			newAddrs = append(newAddrs, &Update{Addr: a + ":" + strconv.Itoa(int(s.Port)),
+				Metadata: AddrMetadataGRPCLB{AddrType: GRPCLB, ServerName: s.Target}})
+		}
+	}
+	return newAddrs
+}
+
+func (w *dnsWatcher) lookupHost() []*Update {
+	var newAddrs []*Update
+	addrs, err := lookupHost(w.ctx, w.host)
+	if err != nil {
+		grpclog.Warningf("grpc: failed dns A record lookup due to %v.\n", err)
+		return nil
+	}
+	for _, a := range addrs {
+		a, ok := formatIP(a)
+		if !ok {
+			grpclog.Errorf("grpc: failed IP parsing due to %v.\n", err)
+			continue
+		}
+		newAddrs = append(newAddrs, &Update{Addr: a + ":" + w.port})
+	}
+	return newAddrs
+}
+
+func (w *dnsWatcher) lookup() []*Update {
+	newAddrs := w.lookupSRV()
+	if newAddrs == nil {
+		// If failed to get any balancer address (either no corresponding SRV for the
+		// target, or caused by failure during resolution/parsing of the balancer target),
+		// return any A record info available.
+		newAddrs = w.lookupHost()
+	}
+	result := w.compileUpdate(newAddrs)
+	w.curAddrs = newAddrs
+	// Next lookup should happen after an interval defined by w.r.freq.
+	w.t.Reset(w.r.freq)
```

Comment:
we cannot move the timer blocking inside lookup since then we also need to move the ctx inside lookup. I'll move out the reset logic for clarity.